File: apps/api/src/quant_etf_api/schemas/industry.py
"""申万行业轮动子系统 API 模型（调试页与只读研究端点）。"""

from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field

from quant_etf_api.schemas.types import UtcDatetime


class IndustryIndexSummary(BaseModel):
    """申万一级行业摘要。"""

    industry_code: str
    name_cn: str
    is_benchmark_excluded: bool = False


class IndustryRRGPoint(BaseModel):
    """单行业单日 RRG 值。"""

    trade_date: str
    industry_code: str
    name_cn: str = ""
    rs_ratio: float | None = None
    rs_momentum: float | None = None
    quadrant: int | None = None


class IndustryLabIssue(BaseModel):
    """调试页数据问题条目（替代“空结果=无数据”的静默语义）。"""

    level: str = "warn"
    code: str
    message: str
    industry_codes: list[str] = Field(default_factory=list)
    count: int | None = None
    sample_dates: list[str] = Field(default_factory=list)


class IndustryRRGCoverageItem(BaseModel):
    """单行业 RRG 数据覆盖与有效区间（按行业日线质量统计）。"""

    industry_code: str
    name_cn: str = ""
    data_start_date: date | None = None
    data_end_date: date | None = None
    expected_input_days: int = 0
    present_input_days: int = 0
    missing_input_days: int = 0
    output_days: int = 0
    leading_nan_days: int = 0
    valid_count: int = 0
    valid_from: date | None = None
    valid_until: date | None = None


class IndustryRRGMeta(BaseModel):
    """RRG 查询的计算口径与数据问题汇总。"""

    requested_start: date
    requested_end: date
    effective_start: date | None = None
    effective_end: date | None = None
    market_trading_days: int = 0
    warmup_required: int = 0
    max_range_days: int = 0
    range_exceeded: bool = False
    issues: list[IndustryLabIssue] = Field(default_factory=list)
    coverage: list[IndustryRRGCoverageItem] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)


class IndustryRRGResponse(BaseModel):
    """RRG 序列响应。"""

    points: list[IndustryRRGPoint]
    warmup_days: int = Field(default=0, description="指标 warm-up 交易日数")
    meta: IndustryRRGMeta | None = None


class IndustryIndexCorrelationItem(BaseModel):
    """一个行业与指定指数的日收益相关度。"""

    industry_code: str
    name_cn: str = ""
    correlation: float | None = None
    sample_count: int = 0


class IndustryIndexCorrelationResponse(BaseModel):
    """指定指数与多个行业的相关度查询响应。"""

    index_code: str
    start: date
    end: date
    index_close_days: int = 0
    items: list[IndustryIndexCorrelationItem] = Field(default_factory=list)


class IndustryDiffusionPoint(BaseModel):
    """单行业单日扩散指标值。"""

    trade_date: str
    industry_code: str
    name_cn: str = ""
    value: float | None = None
    valid_count: int | None = None
    member_count: int | None = None
    coverage: float | None = None


class IndustryDiffusionCoverageItem(BaseModel):
    """单行业扩散数据覆盖统计（成员股、样本、无样本日）。"""

    industry_code: str
    name_cn: str = ""
    member_count: int = 0
    data_start_date: date | None = None
    data_end_date: date | None = None
    output_days: int = 0
    no_sample_days: int = 0
    valid_count: int = 0
    valid_from: date | None = None
    valid_until: date | None = None
    avg_sample_count: float | None = None


class IndustryDiffusionMeta(BaseModel):
    """扩散查询的计算口径与数据问题汇总。"""

    requested_start: date
    requested_end: date
    effective_start: date | None = None
    effective_end: date | None = None
    market_trading_days: int = 0
    output_days: int = 0
    max_range_days: int = 0
    range_exceeded: bool = False
    issues: list[IndustryLabIssue] = Field(default_factory=list)
    coverage: list[IndustryDiffusionCoverageItem] = Field(default_factory=list)
    rules: list[str] = Field(default_factory=list)


class IndustryDiffusionResponse(BaseModel):
    """扩散指标序列响应。"""

    points: list[IndustryDiffusionPoint]
    meta: IndustryDiffusionMeta | None = None


class IndustryRotationSelection(BaseModel):
    """单日轮动选择结果。"""

    trade_date: str
    selected_codes: list[str] = Field(default_factory=list)
    weights: dict[str, float] = Field(default_factory=dict)


class IndustryRotationResponse(BaseModel):
    """轮动选择结果响应。"""

    selections: list[IndustryRotationSelection]


class IndustrySummaryItem(BaseModel):
    """行业数据管理列表摘要（元数据 + 成分数 + 质量快照 + 最新行情）。"""

    industry_code: str
    name_cn: str
    is_benchmark_excluded: bool = False
    member_count: int = 0
    data_start_date: date | None = None
    data_end_date: date | None = None
    bar_count: int | None = None
    missing_day_count: int | None = None
    quality_checked_at: UtcDatetime | None = None
    latest_trade_date: date | None = None
    latest_close: float | None = None
    latest_change_pct: float | None = None


class IndustryQualityDetail(BaseModel):
    """行业详情页数据质量（快照 + 日线字段完整性）。"""

    industry_code: str
    data_start_date: date | None = None
    data_end_date: date | None = None
    bar_count: int | None = None
    missing_day_count: int | None = None
    quality_checked_at: UtcDatetime | None = None
    total: int = 0
    min_date: date | None = None
    max_date: date | None = None
    missing_open: int = 0
    missing_high: int = 0
    missing_low: int = 0
    missing_close: int = 0
    incomplete_rows: int = 0
    incomplete_ratio: float = 0.0
    change_pct_null: int = 0
    change_pct_null_rate: float = 0.0

File: apps/web/src/api/industry.ts
import { apiClient } from './client'

/** 申万一级行业摘要 */
export interface IndustryIndexSummary {
  industry_code: string
  name_cn: string
  is_benchmark_excluded: boolean
}

/** 行业数据管理列表摘要（元数据 + 成分数 + 质量快照 + 最新行情） */
export interface IndustrySummaryItem {
  industry_code: string
  name_cn: string
  is_benchmark_excluded: boolean
  member_count: number
  data_start_date: string | null
  data_end_date: string | null
  bar_count: number | null
  missing_day_count: number | null
  quality_checked_at: string | null
  latest_trade_date: string | null
  latest_close: number | null
  latest_change_pct: number | null
}

/** 行业详情页数据质量 */
export interface IndustryQualityDetail {
  industry_code: string
  data_start_date: string | null
  data_end_date: string | null
  bar_count: number | null
  missing_day_count: number | null
  quality_checked_at: string | null
  total: number
  min_date: string | null
  max_date: string | null
  missing_open: number
  missing_high: number
  missing_low: number
  missing_close: number
  incomplete_rows: number
  incomplete_ratio: number
  change_pct_null: number
  change_pct_null_rate: number
}

/** RRG 单点数据 */
export interface IndustryRRGPoint {
  trade_date: string
  industry_code: string
  name_cn: string
  rs_ratio: number | null
  rs_momentum: number | null
  quadrant: number | null
}

/** 调试页数据问题条目 */
export interface IndustryLabIssue {
  level: 'warn' | 'error' | 'info'
  code: string
  message: string
  industry_codes: string[]
  count: number | null
  sample_dates: string[]
}

/** 单行业 RRG 数据覆盖与有效区间 */
export interface IndustryRRGCoverageItem {
  industry_code: string
  name_cn: string
  data_start_date: string | null
  data_end_date: string | null
  expected_input_days: number
  present_input_days: number
  missing_input_days: number
  output_days: number
  leading_nan_days: number
  valid_count: number
  valid_from: string | null
  valid_until: string | null
}

/** RRG 查询元信息 */
export interface IndustryRRGMeta {
  requested_start: string
  requested_end: string
  effective_start: string | null
  effective_end: string | null
  market_trading_days: number
  warmup_required: number
  max_range_days: number
  range_exceeded: boolean
  issues: IndustryLabIssue[]
  coverage: IndustryRRGCoverageItem[]
  notes: string[]
}

/** RRG 响应 */
export interface IndustryRRGResponse {
  points: IndustryRRGPoint[]
  warmup_days: number
  meta: IndustryRRGMeta | null
}

/** 单行业与指定指数的日收益相关度 */
export interface IndustryIndexCorrelationItem {
  industry_code: string
  name_cn: string
  correlation: number | null
  sample_count: number
}

/** 行业与指定指数相关度响应 */
export interface IndustryIndexCorrelationResponse {
  index_code: string
  start: string
  end: string
  index_close_days: number
  items: IndustryIndexCorrelationItem[]
}

/** 指数单日成分扩散调试结果 */
export interface IndexDiffusionDebugResponse {
  index_code: string
  trade_date: string
  factor_value: number | null
  raw_ratio: number | null
  member_count: number
  valid_sample_count: number
  missing_sample_count: number
  rising_sample_count: number
  valid_days: number
  window_complete: boolean
  lookback: number
  smooth_window: number
  calculation_version: string
  calculation_date_count: number
  stock_count: number
  stock_close_point_count: number
}

/** 扩散单点数据 */
export interface IndustryDiffusionPoint {
  trade_date: string
  industry_code: string
  name_cn: string
  value: number | null
  valid_count: number | null
  member_count: number | null
  coverage: number | null
}

/** 单行业扩散数据覆盖统计 */
export interface IndustryDiffusionCoverageItem {
  industry_code: string
  name_cn: string
  member_count: number
  data_start_date: string | null
  data_end_date: string | null
  output_days: number
  no_sample_days: number
  valid_count: number
  valid_from: string | null
  valid_until: string | null
  avg_sample_count: number | null
}

/** 扩散查询元信息 */
export interface IndustryDiffusionMeta {
  requested_start: string
  requested_end: string
  effective_start: string | null
  effective_end: string | null
  market_trading_days: number
  output_days: number
  max_range_days: number
  range_exceeded: boolean
  issues: IndustryLabIssue[]
  coverage: IndustryDiffusionCoverageItem[]
  rules: string[]
}

/** 扩散响应 */
export interface IndustryDiffusionResponse {
  points: IndustryDiffusionPoint[]
  meta: IndustryDiffusionMeta | null
}

/** 轮动单日选择 */
export interface IndustryRotationSelection {
  trade_date: string
  selected_codes: string[]
  weights: Record<string, number>
}

/** 行业轮动查询参数 */
export interface RotationParams {
  signal: 'quadrant' | 'diffusion' | 'diffusion_rrg'
  topN: number
  keepQuadrants: number[]
  lookbackRatio: number
  lookbackMom: number
  smoothWindow: number
  diffusionLookback: number
}

/** 拉取申万一级行业目录 */
export async function fetchIndustryIndexes(): Promise<IndustryIndexSummary[]> {
  const { data } = await apiClient.get<IndustryIndexSummary[]>('/industry/indexes')
  return data
}

/** 拉取行业数据管理列表摘要 */
export async function fetchIndustrySummaries(): Promise<IndustrySummaryItem[]> {
  const { data } = await apiClient.get<IndustrySummaryItem[]>('/industry/summary')
  return data
}

/** 拉取单个申万一级行业日线 K 线数据 */
export async function fetchIndustryDailyBars(
  industryCode: string,
  params: { startDate?: string; endDate?: string; limit?: number } = {},
): Promise<DailyBarLike[]> {
  const { data } = await apiClient.get<DailyBarLike[]>(
    `/industry/indexes/${industryCode}/bars`,
    {
      params: {
        start_date: params.startDate || undefined,
        end_date: params.endDate || undefined,
        limit: params.limit ?? 250,
      },
    },
  )
  return data
}

/** 拉取单个申万一级行业的数据质量详情 */
export async function fetchIndustryQuality(
  industryCode: string,
): Promise<IndustryQualityDetail> {
  const { data } = await apiClient.get<IndustryQualityDetail>(
    `/industry/indexes/${industryCode}/quality`,
  )
  return data
}

/** 行业日线 K 线明细（与系统 DailyBar 一致，本地最小接口避免循环依赖） */
export interface DailyBarLike {
  trade_date: string
  code: string
  open_price: number | null
  high_price: number | null
  low_price: number | null
  close_price: number | null
  prev_close_price: number | null
  change_pct: number | null
  volume: number | null
  turnover: number | null
  source: string
  ingested_at?: string | null
}

/**
 * 拉取 RRG 序列。
 *
 * @param start - 起始日期 YYYY-MM-DD
 * @param end - 截止日期 YYYY-MM-DD
 * @param codes - 行业代码列表
 * @param params - RRG 参数
 * @param options - 可选超时配置；RRG/扩散为研究计算，默认放宽到 120 秒
 * @returns RRG 序列点
 */
export async function fetchRRG(
  start: string,
  end: string,
  codes: string[],
  params: Pick<RotationParams, 'lookbackRatio' | 'lookbackMom' | 'smoothWindow'>,
  options: { timeout?: number } = {},
): Promise<IndustryRRGResponse> {
  const { data } = await apiClient.get<IndustryRRGResponse>('/industry/rrg', {
    params: {
      start,
      end,
      industry_codes: codes.join(','),
      lookback_ratio: params.lookbackRatio,
      lookback_mom: params.lookbackMom,
      smooth_window: params.smoothWindow,
    },
    timeout: options.timeout ?? 120_000,
  })
  return data
}

/**
 * 拉取指定指数与行业的严格日收益相关度。
 *
 * @param indexCode - 指数代码
 * @param start - 起始日期 YYYY-MM-DD
 * @param end - 截止日期 YYYY-MM-DD
 * @param codes - 行业代码列表
 * @returns 按相关度降序排列的行业相关度结果
 */
export async function fetchIndustryIndexCorrelation(
  indexCode: string,
  start: string,
  end: string,
  codes: string[],
): Promise<IndustryIndexCorrelationResponse> {
  const { data } = await apiClient.get<IndustryIndexCorrelationResponse>('/industry/correlation', {
    params: {
      index_code: indexCode,
      start,
      end,
      industry_codes: codes.join(','),
    },
  })
  return data
}

/**
 * 即时计算指定指数在一个交易日的严格成分扩散结果。
 *
 * 仅用于研究调试，不写入因子值表。
 *
 * @param indexCode - 指数代码
 * @param tradeDate - 目标交易日 YYYY-MM-DD
 * @returns 原始上涨占比、平滑因子值及样本诊断
 */
export async function fetchIndexDiffusionDebug(
  indexCode: string,
  tradeDate: string,
): Promise<IndexDiffusionDebugResponse> {
  const { data } = await apiClient.get<IndexDiffusionDebugResponse>('/industry/index-diffusion', {
    params: { index_code: indexCode, trade_date: tradeDate },
    timeout: 120_000,
  })
  return data
}

/**
 * 拉取数量占比扩散序列。
 *
 * @param start - 起始日期 YYYY-MM-DD
 * @param end - 截止日期 YYYY-MM-DD
 * @param codes - 行业代码列表
 * @param params - 扩散参数
 * @param withCoverage - 是否逐点返回每日覆盖度
 * @param options - 可选超时配置；扩散为研究计算，默认放宽到 120 秒
 * @returns 扩散序列点
 */
export async function fetchDiffusion(
  start: string,
  end: string,
  codes: string[],
  params: Pick<RotationParams, 'diffusionLookback' | 'smoothWindow'>,
  withCoverage = true,
  options: { timeout?: number } = {},
): Promise<IndustryDiffusionResponse> {
  const { data } = await apiClient.get<IndustryDiffusionResponse>('/industry/diffusion', {
    params: {
      start,
      end,
      industry_codes: codes.join(','),
      diffusion_lookback: params.diffusionLookback,
      smooth_window: params.smoothWindow,
      with_coverage: withCoverage,
    },
    timeout: options.timeout ?? 120_000,
  })
  return data
}

/**
 * 拉取轮动选择（默认月末决策）。
 *
 * @param start - 起始日期 YYYY-MM-DD
 * @param end - 截止日期 YYYY-MM-DD
 * @param codes - 行业代码列表
 * @param params - 轮动参数
 * @returns 逐决策日选择结果
 */
export async function fetchRotationSelections(
  start: string,
  end: string,
  codes: string[],
  params: RotationParams,
): Promise<IndustryRotationSelection[]> {
  const { data } = await apiClient.get<{ selections: IndustryRotationSelection[] }>(
    '/industry/rotation',
    {
      params: {
        start,
        end,
        industry_codes: codes.join(','),
        signal: params.signal,
        top_n: params.topN,
        keep_quadrants: params.keepQuadrants.join(','),
        lookback_ratio: params.lookbackRatio,
        lookback_mom: params.lookbackMom,
        smooth_window: params.smoothWindow,
        diffusion_lookback: params.diffusionLookback,
        monthly: true,
      },
    },
  )
  return data.selections
}
